File: 01-machine/05-blocking-vs-nonblocking-io/nodejs/io_demo.js
// Layer 1 - Blocking vs non-blocking IO. Node's net sockets are ALWAYS
// non-blocking/event-driven under the hood (libuv registers them with
// epoll/kqueue for you; there is no synchronous socket API in normal use).
// So the "serial" case below isn't OS-level blocking IO the way Python's
// or Go's is -- it's US choosing to await one request before starting the
// next. That distinction is worth sitting with: in Node, "serial vs
// concurrent" is a decision your code makes about *scheduling*, not
// something the IO layer forces on you the way a blocking recv() does.

const net = require("net");

const RESPONSE_DELAY = 100;
const N = 20;

function startServer() {
  return new Promise((resolve) => {
    const server = net.createServer((socket) => {
      socket.on("data", () => {
        setTimeout(() => socket.end("ok"), RESPONSE_DELAY);
      });
    });
    server.listen(0, "127.0.0.1", () => resolve({ server, port: server.address().port }));
  });
}

function request(port) {
  return new Promise((resolve, reject) => {
    const socket = net.createConnection(port, "127.0.0.1", () => {
      socket.write("ping");
    });
    socket.on("data", () => {
      socket.end();
      resolve();
    });
    socket.on("error", reject);
  });
}

async function benchSerial(port) {
  const start = Date.now();
  for (let i = 0; i < N; i++) {
    await request(port); // deliberately not starting the next until this resolves
  }
  return (Date.now() - start) / 1000;
}

async function benchConcurrent(port) {
  const start = Date.now();
  await Promise.all(Array.from({ length: N }, () => request(port)));
  return (Date.now() - start) / 1000;
}

(async () => {
  const { server, port } = await startServer();
  const tSerial = await benchSerial(port);
  const tConcurrent = await benchConcurrent(port);
  console.log(`N=${N} requests, ${RESPONSE_DELAY}ms server delay each`);
  console.log(`"serial" (awaited one at a time): ${tSerial.toFixed(3)}s  (~${((tSerial / N) * 1000).toFixed(0)}ms/req)`);
  console.log(`concurrent (Promise.all):         ${tConcurrent.toFixed(3)}s`);
  server.close();
})();

File: 01-machine/05-blocking-vs-nonblocking-io/golang/io_demo.go
// Layer 1 - Blocking vs non-blocking IO.
// net.Conn.Read/Write present a blocking-looking API to your goroutine --
// the call appears to just wait until data is ready, the way a classic
// blocking socket would. Underneath, Go's runtime registers the file
// descriptor with its netpoller (epoll on Linux) and parks ONLY the
// goroutine, not an OS thread, until the fd is ready -- then resumes it.
// That's why "serial" below still doesn't tie up a full OS thread per
// waiting request the way a truly blocking read(2) call in C would; it's
// a middle ground between Python's raw blocking sockets and Node's
// explicit event-driven API, and it's a big part of why Go code reads like
// synchronous code but scales like async code.
package main

import (
	"fmt"
	"net"
	"sync"
	"time"
)

const ResponseDelay = 100 * time.Millisecond
const N = 20

func startServer() net.Listener {
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		panic(err)
	}
	go func() {
		for {
			conn, err := ln.Accept()
			if err != nil {
				return
			}
			go handleConn(conn)
		}
	}()
	return ln
}

func handleConn(conn net.Conn) {
	defer conn.Close()
	buf := make([]byte, 1024)
	conn.Read(buf)
	time.Sleep(ResponseDelay)
	conn.Write([]byte("ok"))
}

func doRequest(addr string) {
	conn, err := net.Dial("tcp", addr)
	if err != nil {
		panic(err)
	}
	defer conn.Close()
	conn.Write([]byte("ping"))
	buf := make([]byte, 1024)
	conn.Read(buf)
}

func benchSerial(addr string) time.Duration {
	start := time.Now()
	for i := 0; i < N; i++ {
		doRequest(addr)
	}
	return time.Since(start)
}

func benchConcurrent(addr string) time.Duration {
	start := time.Now()
	var wg sync.WaitGroup
	for i := 0; i < N; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			doRequest(addr)
		}()
	}
	wg.Wait()
	return time.Since(start)
}

func main() {
	ln := startServer()
	defer ln.Close()
	addr := ln.Addr().String()

	tSerial := benchSerial(addr)
	tConcurrent := benchConcurrent(addr)
	fmt.Printf("N=%d requests, %v server delay each\n", N, ResponseDelay)
	fmt.Printf("serial (net.Dial, one goroutine):   %v  (~%v/req)\n", tSerial, tSerial/N)
	fmt.Printf("concurrent (goroutine per request):  %v\n", tConcurrent)
}
